File: Editor/src/com/kotcrab/vis/editor/util/CameraZoomController.java
/*
 * Copyright 2014-2015 Pawel Pastuszak
 *
 * This file is part of VisEditor.
 *
 * VisEditor is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * VisEditor is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with VisEditor.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.kotcrab.vis.editor.util;

import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.math.Vector3;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.kotcrab.vis.editor.module.ModuleInput;

public class CameraZoomController {
	private Vector3 unprojectVec;
	private final OrthographicCamera camera;

	public CameraZoomController (OrthographicCamera camera, Vector3 unprojectVec) {
		this.camera = camera;
		this.unprojectVec = unprojectVec;
	}

	/**
	 * Zooms camera around given point
	 * @param x screen x
	 * @param y screen y
	 * @param amount from {@link ModuleInput#scrolled(InputEvent, float, float, int)}
	 */
	public boolean zoomAroundPoint (float x, float y, int amount) {
		float newZoom = 0;
		camera.unproject(unprojectVec.set(x, y, 0));
		float cursorX = unprojectVec.x;
		float cursorY = unprojectVec.y;

		if (amount == -1) { // zoom in
			if (camera.zoom <= 0.3f) return false;
			newZoom = camera.zoom - 0.1f * camera.zoom * 2;

		}

		if (amount == 1) { // zoom out
			if (camera.zoom >= 10f) return false;
			newZoom = camera.zoom + 0.1f * camera.zoom * 2;
		}

		// some complicated calculations, basically we want to zoom in/out where mouse pointer is
		camera.position.x = cursorX + (newZoom / camera.zoom) * (camera.position.x - cursorX);
		camera.position.y = cursorY + (newZoom / camera.zoom) * (camera.position.y - cursorY);
		camera.zoom = newZoom;

		return true;
	}
}

File: editor/src/main/kotlin/com/kotcrab/vis/editor/util/gdx/CameraZoomController.kt
/*
 * Copyright 2014-2016 See AUTHORS file.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.kotcrab.vis.editor.util.gdx

import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.math.Vector3

/**
 * Camera controller that supports zooming around point.
 * @author Kotcrab
 */
class CameraZoomController(private val camera: OrthographicCamera, private val unprojectVec: Vector3) {
    private companion object {
        val MIN_ZOOM = 0.3f
        val MAX_ZOOM = 15f
    }

    /**
     * Zooms camera around given point.
     * @param x screen x
     * @param y screen y
     * @param zoomIn if true then zoomIn will be performed
     */
    fun zoomAroundPoint(x: Float, y: Float, zoomIn: Boolean): Boolean {
        camera.unproject(unprojectVec.set(x, y, 0f))
        val cursorX = unprojectVec.x
        val cursorY = unprojectVec.y

        val (zoomPossible, newZoom) = calcNewZoom(zoomIn)
        if (zoomPossible == false) return false

        // some complicated calculations, basically we want to zoom in/out where mouse pointer is
        camera.position.x = cursorX + newZoom / camera.zoom * (camera.position.x - cursorX)
        camera.position.y = cursorY + newZoom / camera.zoom * (camera.position.y - cursorY)
        camera.zoom = newZoom
        return true
    }

    /** Zooms camera in. */
    fun zoomIn(): Boolean = zoom(true)

    /** Zooms camera out. */
    fun zoomOut(): Boolean = zoom(false)

    private fun zoom(zoomIn: Boolean): Boolean {
        val (zoomPossible, newZoom) = calcNewZoom(zoomIn)
        if (zoomPossible == false) return false
        camera.zoom = newZoom
        return true
    }

    private fun calcNewZoom(zoomIn: Boolean): Pair<Boolean, Float> {
        val newZoom: Float
        if (zoomIn) {
            if (camera.zoom <= MIN_ZOOM) return Pair(false, camera.zoom)
            newZoom = camera.zoom - 0.1f * camera.zoom * 2f
        } else {
            if (camera.zoom >= MAX_ZOOM) return Pair(false, camera.zoom)
            newZoom = camera.zoom + 0.1f * camera.zoom * 2f
        }
        return Pair(true, newZoom)
    }
}
